esp-hal/src/clock/mod.rs

```diff
@@ -602,65 +610,131 @@ impl Clocks {
+/// As long as at least one [PhyClockGuard] exists, the PHY clock will remain
+/// active. To release this guard, you can either let it go out of scope or use
+/// [PhyClockGuard::release] to explicitly release it.
+pub struct PhyClockGuard<'d> {
+    _phantom: PhantomData<&'d ()>,
+}
 #[cfg(any(bt, ieee802154, wifi))]
-impl<'d> RadioClockController<'d> {
-    /// Create a new instance of the radio clock controller
+impl PhyClockGuard<'_> {
     #[instability::unstable]
-    pub fn new(rcc: crate::peripherals::RADIO_CLK<'d>) -> Self {
-        Self { _rcc: rcc }
+    #[inline]
+    /// Release the clock guard.
+    ///
+    /// The PHY clock will be disabled, if this is the last clock guard.
+    pub fn release(self) {}
+}
+#[cfg(any(bt, ieee802154, wifi))]
+impl Drop for PhyClockGuard<'_> {
+    fn drop(&mut self) {
+        decrease_phy_clock_ref_count_internal();
     }
+}
+#[cfg(any(bt, ieee802154, wifi))]
+#[instability::unstable]
+/// This trait provides common functionality for all
+pub trait ModemClockController<'d>: Sealed + 'd {
+    /// Enable the modem clock for this controller.
+    fn enable_modem_clock(&mut self, enable: bool);
 
-    /// Enable the PHY clocks
-    #[instability::unstable]
-    #[cfg(phy)]
-    #[inline]
-    pub fn enable_phy(&mut self, enable: bool) {
-        clocks_ll::enable_phy(enable);
+    /// Enable the PHY clock and acquire a [PhyClockGuard].
+    ///
+    /// The PHY clock will only be disabled, once all [PhyClockGuard]'s of all
+    /// modems were dropped.
+    fn enable_phy_clock(&self) -> PhyClockGuard<'d> {
+        increase_phy_clock_ref_count_internal();
+        PhyClockGuard {
+            _phantom: PhantomData,
+        }
     }
 
-    /// Enable the Bluetooth clocks
-    #[instability::unstable]
-    #[cfg(bt)]
-    #[inline]
-    pub fn enable_bt(&mut self, enable: bool) {
-        clocks_ll::enable_bt(enable);
+    /// Decreases the PHY clock reference count for this modem ignoring
+    /// currently alive [PhyClockGuard]s.
+    ///
+    /// # Panics
+    /// This function panics if the PHY clock is inactive. If the ref count is
+    /// lower than the number of alive [PhyClockGuard]s, dropping a guard can
+    /// now panic.
+    fn decrease_phy_clock_ref_count(&self) {
+        decrease_phy_clock_ref_count_internal();
     }
+}
 
-    /// Enable the WiFi clocks
-    #[instability::unstable]
-    #[cfg(wifi)]
-    #[inline]
-    pub fn enable_wifi(&mut self, enable: bool) {
+#[cfg(wifi)]
+#[instability::unstable]
+impl<'d> ModemClockController<'d> for WIFI<'d> {
+    fn enable_modem_clock(&mut self, enable: bool) {
         clocks_ll::enable_wifi(enable);
     }
-
-    /// Enable the IEEE 802.15.4 peripheral clocks
+}
+#[cfg(wifi)]
+impl WIFI<'_> {
     #[instability::unstable]
-    #[cfg(ieee802154)]
-    #[inline]
-    pub fn enable_ieee802154(&mut self, enable: bool) {
-        clocks_ll::enable_ieee802154(enable);
+    /// Reset the Wi-Fi MAC.
+    pub fn reset_wifi_mac(&mut self) {
```

Comment:
A bit more nit picking, sorry, but these don't need to take `&mut self` - only a single peripheral singleton should exist at a time, so every reference is technically exclusive.

Reply:
🥲 i'll fix it, once I'm back at my laptop

Reply:
actually, let's just keep this as is, the peripehrals are `Sync` currently and accepting `&self` would actually be a problem.